sdk/pinocchio/src/entrypoint/mod.rs

```diff
@@ -166,61 +179,71 @@ macro_rules! program_entrypoint {
-#[allow(clippy::cast_ptr_alignment)]
 #[inline(always)]
-pub unsafe fn deserialize<'a, const ACCOUNTS: usize>(
+pub unsafe fn deserialize(
     mut input: *mut u8,
-    accounts: &mut [core::mem::MaybeUninit<AccountInfo>; ACCOUNTS],
-) -> (&'a Pubkey, usize, &'a [u8]) {
-    // Ensure that the number of accounts is equal to `MAX_TX_ACCOUNTS`.
-    const {
-        assert!(
-            ACCOUNTS == crate::MAX_TX_ACCOUNTS,
-            "The number of accounts must be equal to MAX_TX_ACCOUNTS"
-        );
-    }
-    // The runtime guarantees that the number of accounts does not exceed
-    // `MAX_TX_ACCOUNTS`.
-    let processed = *(input as *const u64) as usize;
-    input = input.add(core::mem::size_of::<u64>());
-
-    for i in 0..processed {
-        let account_info: *mut Account = input as *mut Account;
-        // Adds an 8-bytes offset for:
-        //   - rent epoch in case of a non-duplicated account
-        //   - duplicated marker + 7 bytes of padding in case of a duplicated account
-        input = input.add(core::mem::size_of::<u64>());
-
-        let account = if (*account_info).borrow_state == NON_DUP_MARKER {
-            // Unique account: create a new `AccountInfo` to represent the account.
+    mut accounts: *mut AccountInfo,
+) -> (&'static Pubkey, &'static [u8]) {
+    // Number of accounts to process.
+    let mut to_process = *(input as *const u64) as usize;
+    // Skip the number of accounts (8 bytes).
+    input = input.add(size_of::<u64>());
+
+    if to_process > 0 {
+        // Represents the beginning of the accounts slice.
+        let accounts_slice = accounts;
+
+        // The first account is always non-duplicated, so process
+        // it directly as such.
+        let account: *mut Account = input as *mut Account;
+        accounts.write(AccountInfo { raw: account });
+
+        input = input.add(STATIC_ACCOUNT_DATA + size_of::<u64>());
+        input = input.add((*account).data_len as usize);
+        input = align_pointer!(input);
+
+        to_process -= 1;
+
+        'remaining: while to_process > 0 {
+            // Increment the accounts pointer to the next account.
+            accounts = accounts.add(1);
+
+            // Read the next account.
+            let account: *mut Account = input as *mut Account;
+            // Adds an 8-bytes offset for:
+            //   - rent epoch in case of a non-duplicated account
+            //   - duplicated marker + 7 bytes of padding in case of a duplicated account
+            input = input.add(size_of::<u64>());
+
+            if (*account).borrow_state != NON_DUP_MARKER {
+                accounts.write(AccountInfo {
+                    raw: accounts_slice.add((*account).borrow_state as usize) as *mut Account,
+                });
+                to_process -= 1;
+                // Return to the loop to process the next account (if there is one).
+                continue 'remaining;
+            }
```

Comment:
nit: maybe it's just me, but I find it more legible to avoid the `continue` statement, ie
```suggestion
            if (*account).borrow_state != NON_DUP_MARKER {
                accounts.write(AccountInfo {
                    raw: accounts_slice.add((*account).borrow_state as usize) as *mut Account,
                });
            } else {
              // rest of logic
            }
            to_process -= 1;
```

Reply:
What if we copy the `to_process -= 1;` line into both branches? That would remove the `continue` statement and should give exactly the same performance, right? If not, please add a comment explaining the code :smile:

Reply:
Almost the same, there is still one extra jump on the generated assembly. Perhaps with a later version of the platform tools that Lucas is working on, this code generation is improved.

Reply:
Bah, that stinks, ah well. Can you add a comment explaining the code then?